src/utilities/BatchTableHierarchyExtension.js

```diff
@@ -0,0 +1,199 @@
+import { parseBinArray } from './FeatureTable.js';
+
+function parseProperty( property, buffer, binOffset, length ) {
+
+	if ( ! ( property instanceof Object ) ) {
+
+		throw new Error( 'BatchTableHierarchyExtension: unexpected value found in extension properties.' );
+
+	} else if ( Array.isArray( property ) ) {
+
+		return property;
+
+	} else {
+
+		const byteOffset = property.byteOffset;
+		const componentType = property.componentType || 'UNSIGNED_SHORT';
+
+		const arrayStart = binOffset + byteOffset;
+
+		return parseBinArray( buffer, arrayStart, length, 'SCALAR', componentType );
+
+	}
+
+}
+
+export class BatchTableHierarchyExtension {
+
+	constructor( batchTable ) {
+
+		this.batchTable = batchTable;
+		this.extensionHeader = batchTable.header.extensions[ '3DTILES_batch_table_hierarchy' ];
+
+		this.instancesLength = this.extensionHeader.instancesLength;
+
+		this._cache = {};
+
+	}
+
+	get classIds() {
+
+		if ( ! this._cache.classIds ) {
+
+			const { buffer, binOffset } = this.batchTable;
+			this._cache.classIds = parseProperty(
+				this.extensionHeader.classIds,
+				buffer,
+				binOffset,
+				this.instancesLength,
+			);
+
+		}
+		return this._cache.classIds;
+
+	}
+
+	get parentCounts() {
+
+		if ( ! this._cache.parentCounts ) {
+
+			if ( ! this.extensionHeader.parentCounts ) {
+
+				this._cache.parentCounts = new Array( this.instancesLength );
+				this._cache.parentCounts.fill( 1 );
+
+			} else {
+
+				const { buffer, binOffset } = this.batchTable;
+				this._cache.parentCounts = parseProperty(
+					this.extensionHeader.parentCounts,
+					buffer,
+					binOffset,
+					this.instancesLength,
+				);
+
+			}
+
+		}
+		return this._cache.parentCounts;
+
+	}
+
+	get parentIds() {
+
+		if ( ! this._cache.parentIds ) {
+
+			const { buffer, binOffset } = this.batchTable;
+			this._cache.parentIds = parseProperty(
+				this.extensionHeader.parentIds,
+				buffer,
+				binOffset,
+			);
+
+		}
+		return this._cache.parentIds;
+
+	}
+
+	get classes() {
+
+		if ( ! this._cache.classes ) {
+
+			this._cache.classes = this.extensionHeader.classes;
+
+			const { buffer, binOffset } = this.batchTable;
+			const instances = this._cache.classes.instances;
+
+			for ( const property in instances ) {
+
+				this._cache.classes.instances[ property ] = parseProperty(
+					instances[ property ],
+					buffer,
+					binOffset,
+					this._cache.classes.length,
+				);
+
+			}
+
+		}
+		return this._cache.classes;
+
+	}
+
+	get instancesIds() {
+
+		if ( ! this._cache.instancesIds ) {
+
+			this._cache.instancesIds = [];
+
+			const classCounter = {};
+			for ( const classId of this.classIds ) {
+
+				classCounter[ classId ] = classCounter[ classId ] ?? 0;
+				this._cache.instancesIds.push( classCounter[ classId ] );
+				classCounter[ classId ] ++;
+
+			}
+
+		}
```

Comment:
I opted for the lazy generation so that instancing a `BatchTableHierarchyExtension` does almost nothing but storing the raw data. It saves user who does not care for the batch table in his/her data the cost of parsing batch table hierarchy data, however small that cost. This is inspired by what's done in `FeatureTable` and `BatchTable`, where data is only parsed when `getData` method is called.

I admit it complicates `BatchTableHierarchyExtension` code, with the cache logic. If you would rather have data parsed in constructor, I can change it accordingly.

Reply:
I think we should opt for code simplicity to start and optimize later when needed. The code in `parseBinArray` is very lightweight and I don't think it should cause any issues.

Reply:
Understood, I'll parse data in constructor.